### ISSUE TITLE
Incorrect diagnostics while referencing supertrait associated const

### ISSUE THREAD

**Issue by djc** [rust-lang/rust#72190]
I tried this code:

```rust
trait Trait: SuperTrait {
    fn array() -> [u8; Self::SIZE];
}

trait SuperTrait {
    const SIZE: usize;
}
```

I expected to see this happen: an error that would help me write the correct code.

Instead, this happened: rustc tells me to duplicate the supertrait bound:

```rust
error[E0599]: no associated item named `SIZE` found for type parameter `Self` in the current scope
 --> src/lib.rs:2:30
  |
2 |     fn array() -> [u8; Self::SIZE];
  |                              ^^^^ associated item not found in `Self`
  |
  = help: items from traits can only be used if the type parameter is bounded by the trait
help: the following trait defines an item `SIZE`, perhaps you need to add another supertrait for it:
  |
1 | trait Trait: SuperTrait + SuperTrait {
  |                         ^^^^^^^^^^^^
```

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a60b4876d5d432ebe824d2dbfa93830d

This happens on both stable and nightly.

(I think it should potentially suggest `<Self as SuperTrait>::Size` in this case, but that doesn't seem to work either, see #72192.)

**Comment by estebank** [comment]
What you're trying to write is not currently representable. The diagnostic needs to be fixed as it has for cases inside blocks already.

**Comment by compiler-errors** [comment]
Error message is clear now:

```
error: generic parameters may not be used in const operations
 --> src/lib.rs:2:24
  |
2 |     fn array() -> [u8; Self::SIZE];
  |                        ^^^^^^^^^^ cannot perform const operation using `Self`
  |
  = note: type parameters may not be used in const expressions
  = help: use `#![feature(generic_const_exprs)]` to allow generic const expressions
```

And with `generic_const_exprs`, it just works.
